### ISSUE TITLE
Bug: Save lyrics does not save

### ISSUE THREAD

**Issue by Gero5** [mpogue2/SquareDesk#798]
Here is why I lost my first four files of lyrics editing: After editing the lyrics I pressed the Save-button assuming it would save the lyrics. But it does not. But why is this an active button if it has no function?
greying it out until a file exists would be a solution for this. better: save with the default file name when "save" is clicked by the user.

But I would rather see another more user friendly concept for this. Why do I have "unlock" the lyrics? I think the danger of accidentally changing the lyrics unwillingly is rather small. 
here is what I like it to be: As soon as the user clicks into the yellow area and types or modifies text: the system would auto save it as lyrics for the particular song. If the lyrics file does not exist use the default file name for lyrics <songname>.html and create the file. 
See #796

**Comment by mpogue2** [comment]
If I understand your symptoms correctly, I think there are a couple of bugs here, and they are a use case that I don't think I've ever tested before (which is why there are several bugs here, maybe?).  

Please confirm that the use case below is correct as I've described it!

- You don't have any music files yet, because you just installed, OR you have started the app, but you have not loaded any songs.
- You go to the Lyrics tab, and you see the lyrics template ("Make Some Cuesheets By Morning", etc.), telling you that no cuesheets could be found that match this song (but there isn't a song loaded)
- You click the Unlock for editing button
- That button changes color, but it doesn't really unlock.  <-- BUG number 1 (it's only really unlocked when you see the SAVE and SAVE AS... buttons, and the UNLOCK button disappears)
- You make some edits.
- You click SAVE, because the button is enabled (and it should not be, because there was really no pre-existing lyrics file loaded). <-- BUG number 2
- But the file doesn't actually get saved anywhere. <-- BUG number 3
- If on the other hand you clicked SAVE AS..., the default file name is ".html.html", which does not make sense.  (You'll get a warning that this file will be hidden, because it begins with a dot!  If you click on USE ".", it will actually be saved as ".html.html", but you won't be able to see it unless you use Finder and the CMD-SHIFT-PERIOD trick to enable viewing of hidden files).

Did I get that right?

Here's a short term workaround, until I can address this use case better:
- Make sure that you created the appropriate folders, like "patter", "singing", in your music directory
- Put at least one singing call MP3 file into the "singing" folder
- Quit and restart SquareDesk
- Double click on that singing call to load it
- Go to Lyrics tab, and edit the template.
- Don't use Save (it should be greyed out, if it is just the template), use Save As... for now.
- Save into the Lyrics folder in your music directory
- Then, when you double click on that singing call, it should auto-load the cuesheet (in fact, it will give you the option of loading any of the cuesheets whose names are approximate matches).

Hope this helps!

**Comment by Gero5** [comment]
- **I have about 500 songs in the folders all without lyrics** (is there a public library for sd lyrics?)
- **I select one one of them.** press play and hum the song. (most important information)
- You go to the Lyrics tab, **- yes** and you see the lyrics template **- yes** ("Make Some Cuesheets By Morning", etc.), telling you that no cuesheets could be found that match this song **- yes** (but there isn't a song loaded) **- no. a song is loaded**
- You click the Unlock for editing button **- yes**
- **That button disappears / changes into two buttons "SAVE" and "SAVE AS..."** 
- You make some edits. **- yes**
- You click SAVE, **- yes** because the button is enabled **- yes** (and it should not be, because there was really no pre-existing lyrics file loaded). <-- BUG number 2 **- yes**
- But the file doesn't actually get saved anywhere. <-- BUG number 3 **- yes**
- **and the button changes to UNLOCK FOR EDITING in blue highlighting, wich leaves the impression to me that my file was saved but actually it was not** <-- BUG number ?
- **SAVE AS... is working correctly because of a given song name AFAIS**

- If on the other hand you clicked SAVE AS..., the default file name is ".html.html", which does not make sense. (You'll get a warning that this file will be hidden, because it begins with a dot! If you click on USE ".", it will actually be saved as ".html.html", but you won't be able to see it unless you use Finder and the CMD-SHIFT-PERIOD trick to enable viewing of hidden files). <-- BUG number 4
**- you may be right on that one but that was not MY use case. ;-)**

the workaround works around. thank you.

**Comment by mpogue2** [comment]
re: "is there a public library for sd lyrics?"

Older versions of SquareDesk had a menu item that would automatically download cuesheets that match the files in your music directory.  You might want to see if it's there in version 0.9.5, and try it!

Otherwise, the last time I batched up large numbers of cuesheets, was a long time ago, but you might find this ZIP file useful:

`https://raw.githubusercontent.com/mpogue2/SquareDesk/master/SqViewCueSheets_2017.03.14.zip`

If you unzip it, and stick the "Text" folder into your lyrics folder inside your music directory, SquareDesk will attempt to match them up with the 500 songs you have.

**Comment by mpogue2** [comment]
Yes, the UNLOCK button changes to Blue, which is not intended behavior.  It does not do that if you unlock a cuesheet that is already linked (by SquareDesk) to a music file.  So, this is a use case I'll need to try to debug.

**Comment by Gero5** [comment]
re: Older versions of SquareDesk had a menu item that would automatically download cuesheets that match the files in your music directory. You might want to see if it's there in version 0.9.5, and try it!

Tried. Does nothing for me. my file names are the song name without record label and with some capital letters (Amarillo By Morning.mp3). Maybe the search is missing a key for matching? or my records are just too old... On the other hand "Amarillo By Morning" could find a match if  I interpret the default lyrics sheet right.

**Comment by Gero5** [comment]
re: https://raw.githubusercontent.com/mpogue2/SquareDesk/master/SqViewCueSheets_2017.03.14.zip

Wow. This is huge.

re: If you unzip it, and stick the "Text" folder into your lyrics folder inside your music directory, SquareDesk will attempt to match them up with the 500 songs you have.

I did that while v 0.9.5 was running. it is showing the colored ball for minutes now using up to "100,7" percent of CPU (would you believe that that is possible) due to activity monitor. Killed the process after 10 minutes of waiting. I think the matching algorithm has room for optimization for this use case.
![Bildschirm­foto 2023-03-28 um 10 05 01](https://user-images.githubusercontent.com/128936259/228170886-674412d0-db35-4eb3-ae35-68800810fdf8.jpg)

After that the Lyrics tab has a little asterisk in front of the text and a lot of the lyrics are found. 
Great relief. Awesome. Thank you.

next question: Is there an easy way to identify songs in the list that have no lyrics?

**Comment by Gero5** [comment]
re: Try File > Download Matching Cuesheets....
You mentioned that before. I tried it but it has absolutely no results for me. No cue sheet was found at all. I think it is not working any more. Should I open an issue for that?

re: I don't know why it took >100% of CPU time when the large ZIP file was installed, 
I expanded the zip file first to over 10'000 single files. then renamed the folder to "lyrics". Is it possible that SquareDesk was performing 10'000 update events due to this?

After the restart I find a cue sheet to almost every song that has no typos in the name. That is great. The lyrics folder with the 10'000 cue sheets is 44 MB. it is a big library but concerning disk space I can live with that. Do you think the 10'000 cue sheet are causing performance trouble for SquareDesk?

re: no way to identify songs in the list that have no lyrics. (maybe this could be a feature request?)
No. I don't think so. This would a feature just for me.

**Comment by danlyke** [comment]
We've had some challenges before with the file updates mechanism, so, yeah,
if we got in some pessimal loop of update notifications because of 10k new
files and attempting to re-read the direction ten thousand times, I could
believe that's a problem.

Maybe I'll put a little time into the directory watcher. At the very least,
we should debounce it. This might also be a reason to go to the model-based
table view that we've talked about for other reasons but never made the
leap on.

**Comment by mpogue2** [comment]
@danlyke I believe it's already debounced....I did that a while ago.

**Comment by mpogue2** [comment]
OK, here's why Download Matching Cuesheets does not work right now.

- The cuesheets are stored at SquareDesk.net .
- To figure out which cuesheets are available, we access the website, and get a listing of the files there.
- The format that is used for that listing has changed.
-- It was: `<li><a href="RR%20147%20-%20Amarillo%20By%20Morning.html"> RR 147 - Amarillo By Morning.html</a></li>`
-- It is now: `<a href="RR%20147%20-%20Amarillo%20By%20Morning.html">RR 147 - Amarillo By..&gt;</a>`
- therefore the regex fails, and zero cuesheets are detected.
- Matches all fail, and no cuesheet files are downloaded.

**Comment by mpogue2** [comment]
Download Matching Cuesheets fixed by: 9a30ed67f6505cc567a533808679e21912adfc5c

- updated the regex
- also replaced "%20" by " " (because that's the way the new directory listing encodes the filenames
- I let it scan all my files, and it downloaded >500 matches from the Cloud.  It went pretty fast, and did not over-trigger the FileWatcher.
